File: leetcode_golang/98_validate_bst.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func helper(root *TreeNode, min_val int, max_val int) bool{
    if root == nil {
        return true
    }
    if root.Val <= min_val || root.Val >= max_val {
        return false
    }
    
    return helper(root.Left, min_val, root.Val) && helper(root.Right, root.Val, max_val)
                 
}
func isValidBST(root *TreeNode) bool {
    return helper(root,math.MinInt64, math.MaxInt64)
}

File: leetcode_python/98_validate_bst.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def helper(node, minVal, maxVal):
            if not node:
                return True
            if node.val <= minVal or node.val >= maxVal:
                return False
            return helper(node.left, minVal, node.val) and helper(node.right, node.val, maxVal)
        return helper(root,-sys.maxint-1, sys.maxint)

            
        
